### ISSUE TITLE
Many composes missing from Fedora PDC

### ISSUE THREAD

**Issue by AdamWill** [fedora-infra/pdc-updater#48]
Many of the stable release nightly composes - the `Fedora-Atomic`, `Fedora-Docker`, and `Fedora-Cloud` composes - are missing from PDC. Right now it's 2017-02-17 and there have been composes of all three in the last few days, but the most recent one in PDC is `Fedora-Atomic-25-20170214.0`.

In fact, now I look, we're also missing mainline composes; the most recent Rawhide compose in PDC is `Fedora-Rawhide-20170214.n.0`, but there was a `Fedora-Rawhide-20170215.n.0`...

This is a major problem. PDC has to be reliable and comprehensive.

**Comment by ralphbean** [comment]
- Ah ha.. It looks like pdc-updater is insanely backlogged.  195071 messages in the queue.  I discovered this by running ``/usr/local/bin/fedmsg-service-collectd-fedmsg-hub`` on  ``pdc-backend01``.
- For reasons unknown, the backlog graph for it here does not show up http://threebean.org/fedmsg-health-day.html#pdc-updater
- I get a 500 error from collectd when trying to find more info https://admin.fedoraproject.org/collectd/bin/graph.cgi?hostname=pdc-updater01.phx2.fedoraproject.org;plugin=fedmsg;plugin_instance=hub;type=queue_length;type_instance=PDCUpdater_backlog;begin=-86400
- Here's a better graph found from the collectd UI:  https://admin.fedoraproject.org/collectd/bin/graph.cgi?hostname=pdc-backend01.phx2.fedoraproject.org;plugin=fedmsg;plugin_instance=hub;type=queue_length;type_instance=PDCUpdater_backlog;begin=1484928345;end=1487606745

In short, it looks like those composes will *eventually* be imported.. but the mass rebuild put a bajillion messages in the queue in front of them.

I'll look to see if there's anything I can do to speed that up.

**Comment by ralphbean** [comment]
OK.. this helps explain things.  See this commit in the ansible repo:  https://infrastructure.fedoraproject.org/cgit/ansible.git/commit/?id=97b1e0e3df81c4bd9debd3dbde5d4feeaf7086c4

The pdc-updater backend is limited to use only one thread to handle messages at a time (we usually do 6 or more..).  This makes sense for composes.  Imagine that two composes complete in quick succession.  The data files from them are enormous.  If two threads tried to handle those two composes at the same time, we would OOM (which we did in the past, which is why I set this value to 1).

As of ~Nov pdc-updater now *also* models this dependency graph stuff for a Factory 2.0 goal which means it responds to every *build* that gets *tagged* in koji.  We *should* handle those in parallel, because they're cheap and small (except for texlive).

The solution (I think) is to split pdc-backend out into *two* VMs.  One for compose-triggered stuff, which can be single threaded and unblocked by other stuff.  And one for build-triggered stuff, which can be multithreaded and get clogged up when it needs to be.

**Comment by ralphbean** [comment]
Filed RFR here:  https://pagure.io/fedora-infrastructure/issue/5801

**Comment by AdamWill** [comment]
Sounds reasonable!

**Comment by ralphbean** [comment]
@AdamWill the nodes have been split out now.

I'm re-running ``pdc-updater-initialize`` on pdc-backend01 which should crawl through all of the existing composes it can find and add any that are missing to PDC.  I'll let it run over night.  Can we check on it tomorrow?

**Comment by ralphbean** [comment]
After running `pdc-updater-initialize` for a while, I think it has picked up on the missing composes.

See https://pdc.fedoraproject.org/rest_api/v1/composes/?compose_id=Fedora-Rawhide-20170215.n.0 for an example.

I'll close this.  @AdamWill, please re-open if it isn't fixed.

**Comment by ralphbean** [comment]
Ah, I see a new issue with the branch that we're hitting:

```python
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/moksha/hub/api/consumer.py", line 191, in _work
    self.consume(message)
  File "/usr/lib/python2.7/site-packages/pdcupdater/consumer.py", line 69, in consume
    pdcupdater.utils.handle_message(pdc, self.handlers, msg)
  File "/usr/lib/python2.7/site-packages/pdcupdater/utils.py", line 470, in handle_message
    handler.handle(client, msg)
  File "/usr/lib/python2.7/site-packages/pdcupdater/handlers/compose.py", line 57, in handle
    self._import_compose(pdc, compose_id, compose_url)
  File "/usr/lib/python2.7/site-packages/pdcupdater/utils.py", line 641, in wrapper
    return function(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/pdcupdater/handlers/compose.py", line 140, in _import_compose
    pdcupdater.utils.ensure_release_exists(pdc, release_id, release)
  File "/usr/lib/python2.7/site-packages/pdcupdater/utils.py", line 97, in ensure_release_exists
    pdc['releases']._(release_payload)
  File "<string>", line 1, in <lambda>
  File "/usr/lib/python2.7/site-packages/beanbag/namespace.py", line 131, in fn
    *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/beanbag/url_v1.py", line 102, in call
    return self.make_request(path, verb, kwargs, body)
  File "/usr/lib/python2.7/site-packages/beanbag/url_v1.py", line 155, in make_request
    "Bad response code: %d, %r %r" % (r.status_code, r.request, r))
BeanBagException: Bad response code: 400, <PreparedRequest [POST]> <Response [400]>
```

**Comment by ralphbean** [comment]
beanbag is frustratingly bad at telling me *why* we sent a [bad request](https://http.cat/400).

I've modified url_v1.py on pdc-backend01 to try and get more debug info out the next time this happens.

The root cause is likely that [i upgraded pdc](https://fedorapeople.org/groups/factory2/sprint-010//threebean-pdc-upgrade.ogv) and the required data to create a new release entry (for f26) has changed.  pdc-updater will need a patch to handle that.  Just need to figure out what the missing expected fields are first.
